### PR TITLE
[DebuggerV2] Add tutorial doc

### PR DESCRIPTION
**Pull request by caisq**
* Motivation for features / changes
  * Add documentation and getting-starting guide for debugger-v2.
* Technical description of changes
  * Add a case-based example that walks through how to use debugger-v2 in
     TensorBoard.
  * It covers API usage of `tf.debugging.experiemntal.enable_dump_debug_info()`
  * The organization of the Debugger V2 GUI in TensorBoard and how to use the
     GUI to discover root cause of the bug.
  * Software / hardware compatibility and API usage tips.

### CODE REVIEW

**Comment by caisq** [comment]
Gentle ping.

File: docs/debugger_v2.md [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+[bfloat16](https://en.wikipedia.org/wiki/Bfloat16_floating-point_format) dtype):
+- DType
+- Rank
+- Total number of elements
+-  A breakdown of the floating-type elements into the following categories:
+   negative finite (`-`), zero (`0`), positive finite (`+`), negative infinity
+   (`-∞`), positive infinity (`+∞`), and `NaN`.
+
+The “FULL_HEALTH” mode is suitable for debugging bugs involving NaN and infinity.
+See below for other supported `tensor_debug_mode`s.
```
Comment (bileschi) [inline docs/debugger_v2.md:90]: nit.  consistency.  tensor_debug_mode here is in backtick quotes, where above it is not.

File: docs/debugger_v2.md [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+
+```sh
+python -m tensorflow.python.debug.examples.v2.debug_mnist_v2 \
+    --dump_dir /tmp/tfdbg2_logdir --dump_tensor_debug_mode FULL_HEALTH
+```
+
+## Starting the Debugger V2 GUI in TensorBoard
+
+Running the program with the debugger instrumentation creates a logdir at
+/tmp/tfdbg2_logdir. We can point start TensorBoard and point at the logdir by:
```
Comment (bileschi) [inline docs/debugger_v2.md:111]: nit: phrasing

File: docs/debugger_v2.md [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+![Debugger V2 full view screenshot](./images/debugger_v2_1_full_view.png)
+
+## Using Debugger V2 GUI to find the root cause of NaNs
+
+The Debugger V2 GUI in TensorBoard is organized into six sections:
+- **Alerts**: This top-left section contains a list of “alert” events detected
+  by the debugger in the debug data generated by the instrumented TensorFlow
+  program. Each alert indicates a certain anomaly that warrants attention. In
+  our case, this section highlights 442 NaN/∞ events with a salient pink-red
```
Comment (bileschi) [inline docs/debugger_v2.md:129]: 442?  The images shows 499?

File: docs/debugger_v2.md [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+- **Graph Structure** (bottom half of the top-middle section), **Source Code**
+  (bottom-left section), and **Stack Trace** (bottom-right section) are
+  initially empty. Their contents will be populated when we interact with the
+  GUI. These three sections will also play important roles in our debugging
+  task.
+
+Having oriented ourselves to the organization of the UI, let’s take the
+following steps to get to the bottom of why the NaNs appeared. First, click the
+**NaN/∞** alert in the Alerts section. This automatically scrolls the list of
+600 graph tensors in the Graph Execution section and focuses on the #67, which
```
Comment (bileschi) [inline docs/debugger_v2.md:152]: the image looks like it should be 88, not 67

File: docs/debugger_v2.md [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+is a tensor named “Log:0” generated by a `Log` (natural logarithm) op. A salient
+red color that highlights a -∞ element among the 1000 elements of the 2D float32
```
Comment (bileschi) [inline docs/debugger_v2.md:154]: grammar nit.  Remove "that" here.

File: docs/debugger_v2.md [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+![Debugger V2: Nan / Infinity alerts and graph execution list](./images/debugger_v2_2_nan_inf_alerts.png)
+
+Why does this `Log` op spit out a -∞? Answering that question requires examining
+the input to the op. Clicking on the name of the tensor (“Log:0”) brings up a
+simple but informative visualization of the `Log` op’s vicinity in its
+TensorFlow graph in the Graph Structure section. Note the top-to-bottom
+direction of information flow. The op itself is shown in the bold in the middle.
+Immediately above it, we can see a Placeholder op provides the one and only input
+to the `Log` op. Where is the tensor generated by this logits Placeholder in the
+Graph Execution list? By using the background color and the “input” labels as
```
Comment (bileschi) [inline docs/debugger_v2.md:171]: background color?  They all look grey in the images.

File: docs/debugger_v2.md [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+visual aids, we can see that the `logits:0` tensor is two rows above the `Log:0`
+tensor, that is, in row #65.
```
Comment (bileschi) [inline docs/debugger_v2.md:173]: s/65/85

File: docs/debugger_v2.md [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+
+![Debugger V2: Graph structure view and tracing to input tensor](./images/debugger_v2_3_graph_input.png)
+
+A more careful look at the numerical breakdown of the “logits:0” tensor in row
+#65 reveals why its consumer `Log:0` produces a -∞: Among the 1000 elements of
```
Comment (bileschi) [inline docs/debugger_v2.md:178]: s/65/85

File: docs/debugger_v2.md [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+We are getting closer to solving the bug, but not quite done yet. In order to
+apply the fix, we need to know where in the Python source code the Log op and
+its Placeholder input originated. Debugger V2 provides first-class support for
+tracing the graph ops and execution events to their source. When we clicked the
+`Log:0` tensor in Graph Executions, the Stack Trace section was populated with
+the original stack trace of the Log op’s creation. The stack trace is somewhat
+large because it includes many frames from TensorFlow’s internal code (e.g.,
+gen_math_ops.py and dumping_callback.py), which we can safely ignore for most
+debugging tasks. The frame of interest is Line 216 of debug_mnist_v2.py (i.e,
+the Python file we’re actually trying to debug). Clicking “Line 204” brings up
```
Comment (bileschi) [inline docs/debugger_v2.md:195]: Line 204 -> Line 216

File: docs/debugger_v2.md [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+Catastrophic events involving [NaN](https://en.wikipedia.org/wiki/NaN)s can
+sometimes occur during a TensorFlow program, crippling the model training
+processes. The root cause of such events are often obscure, especially for
+models of non-trivial size and complexity. To make it easier to debug this type
+of model bugs, TensorFlow 2.3+ provides a specialized GUI tool in TensorBoard called Debugger V2. Here we demonstrate how to use this tool by working through a real bug involving NaNs in a neural network written in TensorFlow 2 (TF2).
+
+The techniques illustrated in this tutorial are applicable to other types of
+debugging activities such as inspecting runtime tensor shapes in complex
+programs. This document focuses on a NaN bug due to the high frequency at which
+this type of bugs appear in user programs.
```
Comment (GalOshri) [inline docs/debugger_v2.md:15]: Nit: can shorten this to "This tutorial focuses on NaNs due to their high frequency of occurrence."

File: docs/debugger_v2.md (hunk 1 above)
Comment (caisq) [reply docs/debugger_v2.md:92]: Done a pass throughout the doc and using backticks around tensor_debug_mode consistently now.

File: docs/debugger_v2.md (hunk 2 above)
Comment (caisq) [reply docs/debugger_v2.md]: Done.

File: docs/debugger_v2.md (hunk 3 above)
Comment (caisq) [reply docs/debugger_v2.md]: Ah - the text was from an older run. It turns out that the number differs slightly depending on whether the TF program runs on CPU or GPU. I made changes here to make the text consistent with the screenshots.

File: docs/debugger_v2.md (hunk 4 above)
Comment (caisq) [reply docs/debugger_v2.md]: This is where the exact order varies slightly from run to run. Made changes to make the text consistent with the screenshots.

File: docs/debugger_v2.md (hunk 5 above)
Comment (caisq) [reply docs/debugger_v2.md]: Done.

File: docs/_book.yaml [hunk 11]
```diff
@@ -35,6 +35,8 @@ upper_tabs:
         path: /tensorboard/tensorboard_in_notebooks
       - title: "TensorBoard Data as DataFrames"
         path: /tensorboard/dataframe_api
+      - title: "Debugger V2"
```
[the remark this answers is not in the document]
Comment (caisq) [reply docs/_book.yaml:38]: I agree that in general we should talking about features in a v1/v2 unified way. But currently, the old Debugger (V1) plugin still exists in TensorBoard. Omitting the "V2" part would be confusing to users. 

Once the old Debugger plugin is deprecated and removed, we can remove the V2 here.

File: docs/debugger_v2.md [hunk 12]
```diff
@@ -0,0 +1,303 @@
+# Debugging TensorFlow 2 Programs Using TensorBoard Debugger V2
```
[the remark this answers is not in the document]
Comment (caisq) [reply docs/debugger_v2.md:1]: 1. See my replay about the "V2" above.
2. Good point about "debugging" being an overloaded term. I change the title to "numerical debugging".

File: docs/debugger_v2.md [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+
+> *NOTE*: tf.debugging.experimental.enable_dump_debug_info() is an experimental
+> API and may be subject to breaking changes in the future.
+
+Catastrophic events involving [NaN](https://en.wikipedia.org/wiki/NaN)s can
+sometimes occur during a TensorFlow program, crippling the model training
+processes. The root cause of such events are often obscure, especially for
+models of non-trivial size and complexity. To make it easier to debug this type
+of model bugs, TensorFlow 2.3+ provides a specialized GUI tool in TensorBoard called Debugger V2. Here we demonstrate how to use this tool by working through a real bug involving NaNs in a neural network written in TensorFlow 2 (TF2).
```
[the remark this answers is not in the document]
Comment (caisq) [reply docs/debugger_v2.md]: Done.

File: docs/debugger_v2.md (hunk 10 above)
Comment (caisq) [reply docs/debugger_v2.md]: Done.

File: docs/debugger_v2.md (hunk 6 above)
Comment (caisq) [reply docs/debugger_v2.md:173]: Ah - I was anticipating an in-flight PR (#3765) when writing this. The PR has now landed and I've updated the screenshot here. Thanks for the catch.

File: docs/debugger_v2.md (hunk 8 above)
Comment (caisq) [reply docs/debugger_v2.md]: Done.

File: docs/debugger_v2.md (hunk 9 above)
Comment (caisq) [reply docs/debugger_v2.md:197]: Done.

**Comment by caisq** [comment]
cc @alextp FYI.
